src/engine/engine_common.c

```diff
@@ -432,13 +425,152 @@ static int lock_clines(struct ocf_request *req,
 	}
 }
 
+static inline int ocf_evict_part_overflow(struct ocf_request *req,
+		const struct ocf_engine_callbacks *engine_cbs, uint32_t clines_to_evict)
+{
+	int lock_status = -OCF_ERR_NO_LOCK;
+	struct ocf_metadata_lock *metadata_lock = &req->cache->metadata.lock;
+
+	ocf_metadata_start_exclusive_access(metadata_lock);
+
+	if (space_managment_evict_do(req->cache, req, clines_to_evict) ==
+			LOOKUP_MISS) {
+		goto unlock;
+	}
+
+	if (!ocf_part_is_enabled(req)) {
+		req->info.mapping_error = 1;
+		goto unlock;
+	}
+
+	lock_status = lock_clines(req, engine_cbs);
+
+unlock:
+	ocf_metadata_end_exclusive_access(metadata_lock);
+
+	return lock_status;
+}
+
+static inline int ocf_prepare_clines_hit(struct ocf_request *req,
+		const struct ocf_engine_callbacks *engine_cbs)
+{
+	int lock_status = -OCF_ERR_NO_LOCK;
+	uint32_t clines_to_evict;
+	int res;
+
+	/* Cachelines are mapped in correct partition */
+	if (!ocf_engine_needs_repart(req)) {
+		res = lock_clines(req, engine_cbs);
+		ocf_req_hash_unlock_rd(req);
+		return res;
+	}
+
+    /* Check whether target partition if full */
+	res = ocf_part_check_space(req, &clines_to_evict);
+	if (res == OCF_PART_HAS_SPACE)
+		lock_status = lock_clines(req, engine_cbs);
+
+	if (res == OCF_PART_IS_DISABLED)
+		req->info.mapping_error = 1;
+
+	ocf_req_hash_unlock_rd(req);
+
+	if (res != OCF_PART_IS_FULL)
+		return lock_status;
+
+	/* Target partition is full. Evict missing number of cache lines */
+	lock_status = ocf_evict_part_overflow(req, engine_cbs, clines_to_evict);
+
+	return lock_status;
+}
+
+static inline int ocf_map_free(struct ocf_request *req,
```

Comment:
I fell that the way code is split between subroutines and ocf_engine_prepare_cachelines is artificial. For example this function assumes it is called with hash bucket read lock acquired and it exits with locks released. Seems to me that one long function would be easier to follow. Let's discuss how/if to introduce new sub-routines.